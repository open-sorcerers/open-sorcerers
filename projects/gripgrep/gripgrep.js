'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var execa = _interopDefault(require('execa'));
var F = _interopDefault(require('fluture'));
var ramda = require('ramda');

var KEYS = Object.freeze({
  CWD: "cwd",
  REGEX: "regex",
  GLOBS: "globs",
  STRING: "string"
});

var NIL = "nil";

var box = function (x) { return [x]; };
var orNil = ramda.propOr(NIL);
var isNil = ramda.equals(NIL);
var KÎ© = ramda.once(function (x) { return function () { return x; }; });
var wrap = ramda.curryN(2, function (ab, x) { return ramda.length(ab) === 2 ? ab[0] + x + ab[1] : ab + x + ab; }
);
var quote = wrap("'");
// eslint-disable-next-line valid-typeof
var expectType = ramda.curryN(2, function (type, xx) { return typeof xx === type; });
var expectWhat = ramda.curryN(
  2,
  function (expected, what) { return ("Expected to be given " + expected + " but instead received (" + (typeof what) + ")"); }
);
var yellUnless = ramda.curryN(3, function (yell, expected, what) {
  if (!expectType(expected, what)) {
    yell(expectWhat(expected, what));
    return false
  }
  return true
});

/*
const {
  after,
  attempt,
  attemptP,
  bimap,
  chain,
  chainRej,
  coalesce,
  encase,
  encaseP,
  mapRej,
  node,
  reject,
  rejectAfter,
  resolve,
  swap
} = F
const UNARY = { attempt, attemptP, resolve, reject, swap }
const BINARY = map(curryN(2), {
  after,
  chain,
  chainRej,
  encase,
  encaseP,
  mapRej,
  node,
  rejectAfter
})
const TERNARY = map(curryN(3), { bimap, coalesce })
export const BACK_TO_THE_OLD_FLUTURE = {
  UNARY,
  BINARY,
  TERNARY
}
*/

var futurizeWithCancel = ramda.curryN(3, function (arity, fn, cancel) { return ramda.curryN(arity, function promise2Future() {
    var this$1 = this;
    var args = [], len = arguments.length;
    while ( len-- ) args[ len ] = arguments[ len ];

    return new F(function (bad, good) {
      fn.apply(this$1, [].concat( args ))
        .catch(bad)
        .then(good);
      return cancel
    })
  }); }
);

var unfuturize = ramda.curryN(2, function (arity, fn) { return ramda.curryN(arity, function future2Promise() {
    var this$1 = this;
    var args = [], len = arguments.length;
    while ( len-- ) args[ len ] = arguments[ len ];

    return new Promise(function (resolve, reject) { return F.fork(reject)(resolve)(fn.apply(this$1, args)); }
    )
  }); }
);

var flag = ramda.curryN(2, function (key, value) { return [key, value]; });
var regexFlag = flag("-e");
var stringFlag = flag("-F");
var globFlag = flag("-g");

var reject = F.reject;

var CALLS = Object.freeze({
  raw: ramda.always(["rg", ["--column", "--line-number", "--color", "never"]]),
  regex: regexFlag,
  string: stringFlag,
  globs: ramda.curry(function (globs, flags) { return ramda.reduce(function (agg, c) { return agg.concat(ramda.pipe(quote, globFlag)(c)); }, flags, globs); }
  )
});

var handleCalls = ramda.curry(function (regex, string, globs) {
  var ref = CALLS.raw();
  var cmd = ref[0];
  var flags = ref[1];
  var otherFlags = regex ? CALLS.regex(regex) : CALLS.string(string);
  var globbed = CALLS.globs(globs, []);
  var allFlags = flags.concat(otherFlags, globbed);
  return [cmd, allFlags]
});

/* istanbul-ignore */
var hardOut = function () { return console.log("quitting...") || process.exit(1); };

// execa fxeca -- (fxeca might be a module on its own someday)
var fxeca = futurizeWithCancel(2, execa, hardOut);

var hasErrorOrIsCancelled = ramda.either(
  ramda.propEq("stderr", ""),
  ramda.propEq("isCancelled", true)
);

var formatResults = ramda.map(
  ramda.pipe(ramda.split(":"), function (ref) {
    var file = ref[0];
    var line = ref[1];
    var column = ref[2];
    var match = ref[3];

    return ({
    file: file,
    line: parseInt(line),
    column: parseInt(column),
    match: match
  });
})
);

var handleCancelledCaseOr = ramda.ifElse(ramda.propEq("isCancelled", true), function () { return reject("Process was killed!"); }
);
var handleGoodCase = ramda.pipe(ramda.prop("stdout"), ramda.split("\n"), formatResults);
var handleError = ramda.pipe(ramda.prop("stderr"), reject);

var futureCall = ramda.curry(function (search, ref) {
  var cwd = ref[0];
  var rx = ref[1];
  var globs = ref[2];

  var ref$1 = handleCalls(rx, search, globs);
  var cmd = ref$1[0];
  var args = ref$1[1];
  return cwd ? fxeca(cmd, args, { cwd: cwd }) : fxeca(cmd, args)
});

var manipulate = ramda.ifElse(
  hasErrorOrIsCancelled,
  handleCancelledCaseOr(handleGoodCase),
  handleError
);

var gripgrep = ramda.curry(function (opts, search) { return ramda.pipe(
    // force object
    function (x) { return x || {}; },
    // force array
    box,
    // grab keys safely
    ramda.ap([
      // ripgrep-js required CWD, but because we're using execa we don't need to
      orNil(KEYS.CWD),
      // TODO: Right now since we're basing our API on ripgrep-js, this is set up as a boolean | regex
      // but we could make this boolean only, and then have the search term be potentially a regex
      ramda.propOr(false, KEYS.REGEX),
      ramda.propOr([], KEYS.GLOBS)
    ]),
    // do stuff with the options now that we have them
    ramda.pipe(
      // search => Future(execa)
      futureCall(search),
      // use map b/c we're changing a monad now
      ramda.map(manipulate)
    )
  )(opts); }
);

module.exports = gripgrep;
