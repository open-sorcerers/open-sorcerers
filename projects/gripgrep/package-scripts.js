module.exports = {
  scripts: {
    build: `rollup -c rollup.config.js`
  }
}
