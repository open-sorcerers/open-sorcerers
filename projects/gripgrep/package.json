{
  "name": "gripgrep",
  "version": "0.0.0",
  "description": "ripgrep for JS, in the future",
  "main": "gripgrep.js",
  "repository": "//github.com/open-sorcerers/open-sorcerers",
  "author": "brekk",
  "license": "MIT",
  "private": false,
  "dependencies": {
    "execa": "^4.0.0",
    "fluture": "^12.0.2",
    "ramda": "^0.26.1",
    "yargs-parser": "^16.1.0"
  },
  "devDependencies": {
    "@babel/plugin-transform-modules-commonjs": "^7.7.5",
    "@rollup/plugin-buble": "^0.21.0",
    "@rollup/plugin-commonjs": "^11.0.0",
    "@rollup/plugin-json": "^4.0.1",
    "@rollup/plugin-node-resolve": "^6.0.0",
    "babel-jest": "^24.9.0",
    "eslint": "^6.8.0",
    "eslint-plugin-babel": "^5.3.0",
    "eslint-plugin-fp": "^2.3.0",
    "eslint-plugin-import": "^2.19.1",
    "eslint-plugin-jest": "^23.2.0",
    "eslint-plugin-jsdoc": "^20.0.0",
    "eslint-plugin-node": "^11.0.0",
    "eslint-plugin-prettier": "^3.1.2",
    "eslint-plugin-ramda": "^2.5.1",
    "eslint-plugin-standard": "^4.0.1",
    "jest": "^24.9.0",
    "rollup": "^1.27.14",
    "rollup-plugin-cli": "^0.1.5",
    "rollup-plugin-progress": "^1.1.1",
    "xtrace": "^0.3.0"
  }
}
