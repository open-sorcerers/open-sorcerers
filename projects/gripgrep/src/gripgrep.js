import execa from "execa"
import F from "fluture"
import {
  prop,
  map,
  split,
  either,
  propEq,
  ap,
  curry,
  ifElse,
  pipe,
  propOr,
  reduce,
  always as K
} from "ramda"

import { box, orNil, quote } from "./utils"
import { futurizeWithCancel } from "./future"
import { KEYS } from "./constants"

import { regexFlag, stringFlag, globFlag } from "./cli"
const { reject } = F

export const CALLS = Object.freeze({
  raw: K([`rg`, [`--column`, `--line-number`, `--color`, `never`]]),
  regex: regexFlag,
  string: stringFlag,
  globs: curry((globs, flags) =>
    reduce((agg, c) => agg.concat(pipe(quote, globFlag)(c)), flags, globs)
  )
})

export const prepFlags = curry((regex, string, globs) => {
  const [cmd, flags] = CALLS.raw()
  const otherFlags = regex ? CALLS.regex(regex) : CALLS.string(string)
  const globbed = CALLS.globs(globs, [])
  const allFlags = flags.concat(otherFlags, globbed)
  return [cmd, allFlags]
})

/* istanbul ignore next */
const hardOut = () => console.log("quitting...") || process.exit(1)

// execa fxeca -- (fxeca might be a module on its own someday)
export const fxeca = futurizeWithCancel(2, execa, hardOut)

export const hasErrorOrIsCancelled = either(
  propEq("stderr", ""),
  propEq("isCancelled", true)
)

export const formatResults = map(
  pipe(split(":"), ([file, line, column, match]) => ({
    file,
    line: parseInt(line),
    column: parseInt(column),
    match
  }))
)

export const handleCancelledCaseOr = ifElse(propEq("isCancelled", true), () =>
  reject("Process was killed!")
)
export const handleGoodCase = pipe(prop("stdout"), split("\n"), formatResults)
export const handleError = pipe(prop("stderr"), reject)

export const futureCall = curry((search, [cwd, rx, globs]) => {
  const [cmd, args] = prepFlags(rx, search, globs)
  return fxeca(cmd, args, { cwd })
})

export const manipulate = ifElse(
  hasErrorOrIsCancelled,
  handleCancelledCaseOr(handleGoodCase),
  handleError
)

export const gripgrep = curry((opts, search) =>
  pipe(
    // force object
    x => x || {},
    // force array
    box,
    // grab keys safely
    ap([
      propOr(process.cwd(), KEYS.CWD),
      // TODO: Right now since we're basing our API on ripgrep-js, this is set up as a boolean | regex
      // but we could make this boolean only, and then have the search term be potentially a regex
      propOr(false, KEYS.REGEX),
      propOr([], KEYS.GLOBS)
    ]),
    // do stuff with the options now that we have them
    pipe(
      // search => Future(execa)
      futureCall(search),
      // use map b/c we're changing a monad now
      map(manipulate)
    )
  )(opts)
)
