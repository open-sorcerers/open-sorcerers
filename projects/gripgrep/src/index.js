import { gripgrep } from "./gripgrep"
export default gripgrep
