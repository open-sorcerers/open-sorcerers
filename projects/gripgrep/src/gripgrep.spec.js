import F from "fluture"
import { prepFlags, gripgrep, handleCancelledCaseOr } from "./gripgrep"

test("gripgrep", done => {
  const MAGIC = "magic comment for a test"
  const lookupF = gripgrep({}, MAGIC.slice(0, 13))
  F.fork(done)(out => {
    expect(out).toMatchSnapshot()
    done()
  })(lookupF)
})

test.skip("gripgrep - globs", done => {
  const MAGIC = "fluture"
  const lookupF = gripgrep({ cwd: process.cwd(), globs: ["src/*.js"] }, MAGIC)
  F.fork(e => {
    console.log(">>E>E>>E>E", e)
    done(e)
  })(out => {
    expect(out).toEqual({})
    done()
  })(lookupF)
})

test("gripgrep - force object path", done => {
  const FORCE = "force push objects"
  const lookupF = gripgrep(false, FORCE)
  F.fork(done)(out => {
    expect(out).toMatchSnapshot()
    done()
  })(lookupF)
})

test("handleCancelledCaseOr", done => {
  const ff = handleCancelledCaseOr(done)({ isCancelled: true })
  F.fork(e => {
    expect(e).toEqual("Process was killed!")
    done()
  })(done)(ff)
})

test("prepFlags - regex", () => {
  const regex = "cool"
  const string = false
  const globs = []
  expect(prepFlags(regex, string, globs)).toMatchSnapshot()
})
