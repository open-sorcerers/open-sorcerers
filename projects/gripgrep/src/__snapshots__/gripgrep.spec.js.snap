// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gripgrep - force object path 1`] = `
Array [
  Object {
    "column": 18,
    "file": "src/gripgrep.spec.js",
    "line": 26,
    "match": "  const FORCE = \\"force push objects\\"",
  },
  Object {
    "column": 34,
    "file": "src/__snapshots__/gripgrep.spec.js.snap",
    "line": 9,
    "match": "    \\"match\\"",
  },
]
`;

exports[`gripgrep 1`] = `
Array [
  Object {
    "column": 18,
    "file": "src/gripgrep.spec.js",
    "line": 5,
    "match": "  const MAGIC = \\"magic comment for a test\\"",
  },
  Object {
    "column": 34,
    "file": "src/__snapshots__/gripgrep.spec.js.snap",
    "line": 26,
    "match": "    \\"match\\"",
  },
]
`;

exports[`prepFlags - regex 1`] = `
Array [
  "rg",
  Array [
    "--column",
    "--line-number",
    "--color",
    "never",
    "-e",
    "cool",
  ],
]
`;
